core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala

```diff
@@ -409,6 +409,67 @@ class ConnectionQuotasTest {
     verifyConnectionCountOnEveryListener(connectionQuotas, connectionsPerListener)
   }
 
+  @Test
```

Comment:
It would be useful to add a test where we have both per-listener and per IP limit, and verify that it throttles based on which limit is reached first. Something like: 2 IPs, each per IP limit < per-listener limit, but sum of per IP limits > listener limit. So, if you reach limit on one IP, the broker would not throttle the second IP until it reaches per listener limit. Does not have to be exactly this, just need to verify how per IP throttling interacts with per listener throttling.